### PR TITLE
editor: Add language detection for untitled buffers

### LINKED ISSUE #4868: Automatically detect language for pasted code
### Check for existing issues

- [X] Completed

### Describe the feature

Currently, when I paste copied code into a new empty document, it stays as Plain Text.
In VS Code, it'll automatically detect and set the language. Would be nice if Zed supports this too.

### If applicable, add mockups / screenshots to help present your vision of the feature

_No response_

### PR DESCRIPTION
**Pull request by amtoaer**
# Objective

Closes #4868.

Reintroduce automatic language detection for user-created untitled buffers without allowing it to override language choices made by users or owning surfaces.

More details:
+ https://github.com/zed-industries/zed/pull/61201
+ https://github.com/zed-industries/zed/pull/61351

## Solution

VS Code distinguishes between inferred and explicitly selected languages, and only continues automatic detection in the former state. This change introduces a buffer-local opt-in flag with the same purpose. It defaults to disabled and is enabled only for user-created untitled buffers and restored untitled buffers that did not persist a language. Choosing a language through the language picker disables detection, while automatically detected languages remain eligible for later redetection. Buffers whose language is assigned by their owning surface, such as Channel Notes, never opt in.

Rather than adding persisted state or changing the database schema, restoration reuses the existing optional serialized language. An opted-in buffer that remains Plain Text is serialized without a language and resumes detection after restoration. Once automatic detection has selected a language, that language is serialized normally and restoration leaves detection disabled, effectively treating it as an explicit selection. This also matches VS Code's effective default after restoration.

Candidate selection adopts VS Code's confidence-group filtering. Results below the minimum confidence are discarded, and candidate groups are accepted only after a clear confidence gap. The detector then selects the first available Zed language without eagerly loading every candidate. I then added a separate switching rule: Zed keeps the current language unless another candidate has a clear confidence advantage, reducing language changes as model scores fluctuate during editing. Model inference runs off-thread on a bounded sample, and stale results are discarded.

I also evaluated penalizing YAML and using stricter global switching thresholds, but neither reliably resolves the ambiguity between Markdown and YAML. In the reproduction from #61351, Betlang assigns YAML more than 80% confidence during editing and peaks around 90%. This is reasonable from the content alone: `#` denotes a YAML comment, while `-` denotes a sequence item. Adding ordinary prose or a fenced code block shifts the model strongly toward Markdown. A fixed penalty would overfit this example and risk suppressing legitimate YAML detection. Since known-language buffers do not opt in and manual language selection takes precedence, I consider this level of ambiguity acceptable and prefer not to introduce language-specific score tuning.

Model inference itself is fast. As noted in #61201:

> In release builds on Linux with an Intel Core i5-13600KF, the `betlang::detect` call alone typically completes within 3 ms when processing the maximum sampled input.

The main concern is the work surrounding inference. Without debouncing, rapid typing creates a snapshot and launches a background detection task on every keystroke, while the detected language may change before the user has finished typing. This change therefore adds a 200 ms debounce delay, allowing detection to run and update the language only after the user pauses.

## Testing

Automated coverage verifies that short inputs remain Plain Text, sufficient Rust content is detected as Rust, replacing it with Go content triggers redetection, and manually selecting Plain Text prevents subsequent automatic changes. Restoration coverage verifies that a stored language keeps detection disabled, while an empty language-less untitled buffer resumes detection.

Beyond the automated tests, I manually tested automatic language detection while editing untitled buffers.

## Self-Review Checklist:

- [x] I've reviewed my own diff for quality, security, and reliability
- [x] Unsafe blocks (if any) have justifying comments
- [x] The content adheres to Zed's UI standards ([UX/UI](https://github.com/zed-industries/zed/blob/main/CONTRIBUTING.md#uiux-checklist) and [icon](https://github.com/zed-industries/zed/blob/main/crates/icons/README.md) guidelines)
- [x] Tests cover the new/changed behavior
- [x] Performance impact has been considered and is acceptable

## Showcase

https://github.com/user-attachments/assets/103fc731-e12e-48d6-ae89-03978f1c9d74

---

Release Notes:

- Added automatic language detection for untitled buffers.

### CODE REVIEW

File: crates/editor/src/editor_tests.rs [hunk 1]
```diff
@@ -9603,6 +9602,122 @@ async fn test_kill_ring_yank_pastes_accumulated_kill_at_each_cursor(cx: &mut Tes
...
+        rust_lang().name()
+    );
+
+    cx.executor()
+        .advance_clock(LANGUAGE_DETECTION_DEBOUNCE_TIMEOUT);
+    cx.run_until_parked();
+
+    assert_eq!(
+        buffer.read_with(cx, |buffer, _| buffer.language().unwrap().name()),
+        go_language.name()
```
Comment (MrSubidubi) [inline crates/editor/src/editor_tests.rs:10119]: Curious whether we should really test it like this? As this is still a model at heart, can this not flake in some cases?

Comment (amtoaer) [reply crates/editor/src/editor_tests.rs:10119]: I think this assertion should be stable. Betlang's inference is deterministic, and this input is classified as Go with a clear margin. A future model update could change the prediction, but that would reflect a model behavior change rather than flakiness.

**Comment by ChristopherBiscardi** [comment]
Thanks for taking another go at this!

> Since known-language buffers do not opt in and manual language selection takes precedence, I consider this level of ambiguity acceptable and prefer not to introduce language-specific score tuning.

An important note for review is that the yaml vs markdown issue was occurring in channel notes for staff members. We use channel notes heavily internally so will need to make sure they are set to markdown.

I believe this also caused some flickering related to agent output, but I'll have to dig up the example internally to test the PR with.

**Comment by amtoaer** [comment]
Yes, Channel Notes already explicitly set their buffers to Markdown, so the newly introduced automatic detection is not intended to apply to them. That is why I added a buffer-local opt-in flag and enable it only for user-created untitled buffers, significantly narrowing the scope of the new behavior. If the agent-output flickering has the same cause, I expect this change to address it as well. I’ll test that scenario separately too.

Automatic detection is also disabled when the user manually selects a language. This gives users a way to resolve ambiguous cases even in newly created untitled buffers, which is why I consider occasional misclassifications caused by inherently ambiguous content acceptable.

**Comment by KuSh** [comment]
Perhaps open an issue on betlang for that specific case? At least reducing confidence if it can't detect language correctly?

**Comment by amtoaer** [comment]
That makes sense. I’ll open an issue with the reproduction. Since the snippet is valid as both Markdown and YAML, the YAML prediction itself is understandable; the main concern is that the model assigns it such high confidence despite the ambiguity.

Edit: I opened DioxusLabs/betlang#5 with a minimal reproduction.

**Comment by ealmloff** [comment]
I published a new version of betlang. The new fine tune should be somewhat better here in terms of confidence. It was taking probs and re-softmaxing them on a subset of the classes which skews the confidence incorrectly. The new model should be slightly better here, but since it is ambiguous it will sometimes guess the intent wrong. It also gets a slightly better accuracy in my fs testing. somewhere between 93%-97%

**Comment by ChristopherBiscardi** [comment]
thank you @ealmloff

**Comment by amtoaer** [comment]
I’ve rebased onto `main`, fixed the compilation errors caused by the function signature changes, and updated `betlang` to version 0.1.1. Please let me know if anything else needs to be addressed.

**Comment by ChristopherBiscardi** [comment]
This is mostly good to go at this point. Two more things:

- When saving a file that has previously been auto-detected, it should reset to the appropriate language for the file extension that was chosen. So if you write Rust code, then save that as `something.js`, the file should switch to JavaScript after setting the file's name. (currently it stays in Rust)
- There should be a setting to turn this functionality on/off that isn't the `disable_ai` setting. It is reasonable to expect that someone would want this on or off irrespective of if they have ai on or off.

**Review by ChristopherBiscardi — APPROVED** [review]
Thank you for your work on this!
